actions/test/run_local.ts

```diff
@@ -55,14 +60,92 @@ const main = async () => {
         console.error("[run_local] Error in processBlock", error);
       }
     });
+  } else if (contractAddressEnv) {
+    // If no blockNumberEnv is provided, then we rebuild the state from the default deployment block
+    if (!blockNumberEnv) {
+      console.log(`[run_rebuild] No block number provided, using default deployment block`);
+    } else {
+      assert(!isNaN(Number(blockNumberEnv)), "blockNumber must be a number");
+    }
+    let fromBlock = blockNumberEnv ? Number(blockNumberEnv) : DEFAULT_DEPLOYMENT_BLOCK;
+
+    // Rebuild the state
+    console.log(`[run_rebuild] Rebuild the state of the conditional orders from the historical events.`);
+    const contractAddress = process.env.CONTRACT_ADDRESS;
+    assert(contractAddress && ethers.utils.isAddress(contractAddress), "contract address is required");
+    const pageSize = process.env.PAGE_SIZE ? parseInt(process.env.PAGE_SIZE) : DEFAULT_PAGE_SIZE;
+
+    // 1. Record the current block number - useful with paging
+    let currentBlockNumber = await provider.getBlockNumber();
+    console.log(`[run_rebuild] Current block number: ${currentBlockNumber}`);
+
+    // 2. Connect to the contract instance
+    const contract = ComposableCoW__factory.connect(contractAddress, provider);
+    
+    // 3. Define the filter.
+    const filter = contract.filters.ConditionalOrderCreated();
+
+    // 4. Get the historical events
+    const replayPlan: ReplayPlan = {}
+    let toBlock;
+    do {
+      toBlock = (pageSize == 0) ? 'latest' : fromBlock + (pageSize - 1);
+      if (typeof toBlock !== 'string' && toBlock > currentBlockNumber) {
+          // refresh the current block number
+          currentBlockNumber = await provider.getBlockNumber();
+          toBlock = (toBlock > currentBlockNumber) ? currentBlockNumber : toBlock;
+
+          console.log(`[run_rebuild] Reaching tip of chain, current block number: ${currentBlockNumber}`);
+      }
+
+      console.log(`[run_rebuild] Processing events from block ${fromBlock} to block ${toBlock}`);
+
+      const events = await contract.queryFilter(filter, fromBlock, toBlock);
+
+      console.log(`[run_rebuild] Found ${events.length} events`);
+
+      // 5. Process the events
+      for (const event of events) {
+        if (replayPlan[event.blockNumber] === undefined) {
+          replayPlan[event.blockNumber] = new Set();
+        }
+
+        replayPlan[event.blockNumber].add(event.transactionHash);
+      }
+
+      // only possible string value for toBlock is 'latest'
+      if (typeof toBlock !== 'string') {
+          fromBlock = toBlock + 1;
+      }
+    } while (toBlock !== 'latest' && (typeof toBlock !== 'string' && toBlock != currentBlockNumber));
+
+    // 6. Replay the blocks by iterating over the replayPlan
```

Comment:
One idea to make it more reusable and easier to get the gist is, you can pass the filter as a param, and then pass a async function to process each transaction.

This way, if we wanted, the block indexer can even be unit tested independently, so it can the processing of tx, etc... 

Just a suggesyion, feel free to ignore. If you want to be fancy with this, maybe the block indexer can return an iterator, this way, we just use a more functional approach and we chain the logic. This way, we can start to process TX BEFORE we are done getting all the TX from the blocks.

Reply:
if you want to be even more fancy, you can use generators: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html